File: balanced_binary_tree.py
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
    res=[]
    def printTree(self, root):
        if not root:
            return []
        if root:
            self.res.append(root.val)
            if root.left:
                self.printTree(root.left)
            if root.right:
                self.printTree(root.right)
        return self.res
class Solution:
    def isBalanced(self, root):
        def maxDepth(root):
            if not root:
                return 0
            Left = maxDepth(root.left)
            if Left == -1:
                return -1
            Right = maxDepth(root.right)
            if Right == -1:
                return -1
            return  max(Left,Right)+1 if abs(Left-Right) < 2 else -1
        return maxDepth(root) != -1
if __name__ == '__main__':
    sol = Solution()
    tree = TreeNode(None)
    root= TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.right = TreeNode(4)
    print tree.printTree(root)
    print sol.isBalanced(root)

File: Algorithm/CPP/balanced_binary_tree.cpp
# include <iostream>
# include <string>
using namespace std;



struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
	int getHeight(TreeNode* root)
	{
		if (!root) return 0;
		return 1 + max(getHeight(root->left), getHeight(root->right));
	}
	bool isBalanced(TreeNode* root)
	{
		if (!root) return true;
		if (abs(getHeight(root->left) - getHeight(root->right)) > 1)return false;
		return isBalanced(root->left) && isBalanced(root->right);

	}
	void print(TreeNode* root)
	{
		if (root)
		{
			cout << root->val << " ";
			print(root->left);
			print(root->right);
		}
	}

};

int  main()
{
	Solution sol;
	TreeNode *root1;
	root1 = new TreeNode(5);
	root1->left = new  TreeNode(4);
	root1->right = new TreeNode(8);
	root1->left->left = new TreeNode(11);
	//root1->left->right = new TreeNode(13);
	//root1->right->left = new TreeNode(4);
	root1->left->left->left = new TreeNode(7);
	//root1->left->left->right = new TreeNode(2);
	//root1->right->right->left =new TreeNode(5);
	//root1->right->right->right = new TreeNode(1);
	vector<int> postorder = {2, 3,1};
	vector<int> inorder = { 2, 1, 3 };
	vector<int> nums = { 1, 3, 5, 6, 7 };
	cout << sol.isBalanced(root1) << endl;

	system("pause");
	return 0;
}
